### ISSUE TITLE
connections with and without useMongoClient yield different NativeConnection objects 

### ISSUE THREAD

**Issue by Spown** [Automattic/mongoose#5714]
Mongoose: 4.12.2
OS: Win
Node 7.9.0
MongoDB: 3.2

``NativeConnection`` object from ``onConnected`` event, that was created with and without ``useMongoClient`` option has a slightly different structure. In particular with ``useMongoClient = true`` it has a ``.then()`` method and other promise-related fields, whereas otherwise it doesn't.

It becomes problematic in my case, since I pass it through a Q's promise. Q tries to "intelligently" (which means dumb as a brick, but anyway) resolve the whole promise chain, if it detects one and it does so by testing the resolved value and every chain-descendant value against ``isPromiseLike`` util, which in turn returns true if the said value has the ``.then()`` method.

Now, you can probably see the problem here already. If I depend my app flow on the first promise's triggering - the app would stall if the promise is resolved with such a ``NativeConnection`` object:

```js
var Q = require('q'),
        bootMongo = Q.defer(),
    mongoose = require('mongoose'),
        connOpts = {},
    GITHUB_ISSUE = `gh-5714`
;
mongoose.Promise = require('q').Promise;
connOpts.useMongoClient = true; // toggle for different output
mongoose.connect(`mongodb://localhost:27017/${GITHUB_ISSUE}`, connOpts)
mongoose.connection
.on('error', function(er) {
    bootMongo.reject(new Error(`failed to connect to DB! ${er}`));
})
.on('connected', function() {
    console.log( 'mongooseConnection connected' );
    // for useMongoClient = true ``this`` is a NativeConnection object with .then() method
    // without useMongoClient ``this`` is a NativeConnection object without .then() method
    bootMongo.resolve(this);
})
;
bootMongo.promise
.then(function (conn) {
    // will trigger without useMongoClient = true;
})
.catch(function (err) {
    // will trigger with useMongoClient = true; "promise.then is not a function"
});
```

Technically it is Q who is causing the problem, but Q was doing this in such manner all along and it is also the expected behavior for it. Mongoose however is the one who introduced such game-breaking change.

I can easily fix my problem by wrapping ``this`` from the event callback with another object and pass it into the promise. But I think you still have a problem of inconsistency...

Thank you.

**Comment by vkarpov15** [comment]
Thanks for the detailed repro, will investigate ASAP

**Comment by varunjayaraman** [comment]
@Spown i'm actually getting different results from you. Your promise handlers (the `.then` and `.catch` chained off of `bootMongo` never run for me.

**Comment by Spown** [comment]
``.then`` (with useMongoClient = true) shouldn't run, but are you sure ``.catch`` doesn't run as well?

![](https://drive.google.com/uc?id=0B-OJBofmehO3T0F4UldudHNjX0k)
¯\\(=_=)/¯

**Comment by varunjayaraman** [comment]
sorry, i did something stupid and used `error` instead of `err` inside the `.catch`, and since that error inside the `.catch` was never caught the program seemed to not be doing anything. 

anyway, i was able to reproduce your issue with the following script:

```javascript
var Q = require('q'),
  bootMongo = Q.defer(),
  mongoose = require('mongoose'),
  connOpts = {},
  GITHUB_ISSUE = `gh-5714`;
mongoose.Promise = require('q').Promise;

try {
  connOpts.useMongoClient = true; // toggle for different output
  mongoose.connect(`mongodb://localhost:27017/${GITHUB_ISSUE}`, connOpts);

  mongoose.connection
    .on('error', function(er) {
      console.error(`Error connecting`, er);
      bootMongo.reject(new Error(`failed to connect to DB! ${er}`));
    })
    .on('connected', function() {
      console.log('mongooseConnection connected');
      // for useMongoClient = true ``this`` is a NativeConnection object with .then() method
      // without useMongoClient ``this`` is a NativeConnection object without .then() method

      bootMongo.resolve(this);
    });
  bootMongo.promise
    .then(function(conn) {
      console.log('successfully handled') // never runs 
      // will trigger without useMongoClient = true;
    })
    .catch(function(err) {
      console.error(`Error: ${err}\n${err.stack}`); 
      // will trigger with useMongoClient = true; "promise.then is not a function"
    });
} catch (error) {
  console.error(`Top level error ${error}\n${error.stack}`);
}
```

**Comment by vkarpov15** [comment]
Fixed. Yeah we've been doing a lot to try to make `connect()` work correctly with `async/await` and in order to do that we attach `.then()` and `.catch()` helpers when initial connection is in progress, and then remove those helpers when we're done so we can resolve the promise, because `resolve(this)` causes infinite recursion if `this` has a `.then()`. I added a test case to make sure your case works.
